### DISCUSSION TITLE
snapshot file path construction

### DISCUSSION THREAD

**Discussion by mallison** [simonsmith/cypress-image-snapshot#5] [General]
Hi

I wondered if this change https://github.com/simonsmith/cypress-image-snapshot/blob/master/src/plugin.ts#L67-L69 from the original repository was intentional? It means that the directory that a spec file is in is ignored when constructing the path for the snapshots so all snapshots end up in `<snapshotsDir>/<specFileName>/<snapshotName>.png`.

**Comment by simonsmith** [comment]
Are you referring to the `specFileRelativeToRoot` option from before? I wasn't able to see what this actually did. In my testing I didn't think it made a difference.

Happy to look again if I've overlooked something. Can also look to expand the test cases as well

**Reply by mallison** [reply]
I'm referring to the original implementation here https://github.com/jaredpalmer/cypress-image-snapshot/blob/master/src/plugin.js#L72-L82.

What this does is use the same path under `snapshots` as Cypress uses under `screenshots`.

Eg if Cypress writes the screenshot to 

`/screenshots/path/to/image.png`

 then the corresponding snapshot is written to 

`/snapshots/path/to/image.snap.png`.

The `path/to` part is computed from the path to the spec file.

Your change means that the `path/to` is always discarded, ie the path to the spec file is ignored.

----

The `specFileRelativeToRoot` option is to workaround a change in Cypress 10 with how `path/to` is computed.

Before v10 if you have test at

`cypress/e2e/path/to/test.spec.ts`

then screenshots will be written to

`cypress/screenshots/e2e/path/to/test.spec/screenshot.png`.

What Cypress introduced in v10 is that `path/to` is computed by _removing common path parts_ from the specs being run. This is incredibly annoying because if you run

`cypress/e2e/path/to/test.spec.ts`

with `cypress open` then any snapshots will be written to 

`cypress/snapshots/test.spec/image.snap.png` but if you have, say,

```
cypress/
  e2e/
    path/
      to/
        test.spec.ts
    other/
      test2.spec.ts
```

and run `cypress run` then snapshots will be written to

```
cypress/
  snapshots/
    path/
      to/
        test.spec/
          image.snap.png
    other/
      test2.spec/
        image.snap.png
```

So the file location depends on which tests you decide to run 😱 .

Run `test.spec.ts` in isolation and snapshots are saved to

`cypress/snapshots/test.spec/`.

Run the whole test suite and they are saved to

`cypress/snapshots/path/to/test.spec/`

 I cannot fathom why they made this change.

This is biting me in CI now because I parallelize our Cypress tests so where snapshots are saved depends on which tests are grouped together 😱 .

I guess there are two things here:

1. restore the behaviour where the `snapshots` directory mirrors `screenshots`
2. add an option to override Cypress's silly behaviour and compute snapshot locations from the spec location in a predictable way

Hope that waffle makes sense!

I'm happy to help with both of these.

**Reply by simonsmith** [reply]
@mallison I appreciate your explanation! It makes it clear to me now. I think I can take a look at making a test case and working from that. I agree that this fix should be in here

Once I have taken a closer look I'll push a PR and get your thoughts

**Reply by mallison** [reply]
Great, thanks for taking a look at this!

**Reply by mallison** [reply]
Let me know if I can help out with this.

I'm working on a feature that can only really be tested visually and I've got a few PRs ready I'd love to add snapshot tests to!

**Reply by simonsmith** [reply]
@mallison Sure, if you'd like to have a look at adding this I'd appreciate it! Have been a bit busier than I expected.

**Reply by simonsmith** [reply]
Am actively looking at this again now

**Reply by simonsmith** [reply]
https://github.com/simonsmith/cypress-image-snapshot/pull/8
